python-packaging/execute-code.md

```diff
@@ -0,0 +1,167 @@
+
+### Non-Windows executables
+
+On Linux or Mac systems, the Python file can itself be turned into a command. By adding a [shebang](https://en.wikipedia.org/wiki/Shebang_(Unix))
+as the first line in any Python file, and by giving the file [executable permissions](https://docs.python.org/3/using/unix.html#miscellaneous) the
+file can be directly invoked without a `python` command.
+
+```python
+#!/usr/bin/env python
+# The above line is a shebang, and can take the place of typing python on the command line
+# This comment is below, because shebangs must be the first line!
+
+def shiny_hello():
+    print("\N{Sparkles} Hello from Python \N{Sparkles}")
+
+shiny_hello()
+```
+
+```bash
+my_program.py
+# ✨ Hello from Python ✨
+```
+
+:::{tip}
+Shebangs are a feature of POSIX. POSIX represents some level of compatibility between systems.
+Linux, macOS, all BSDs, and many other operating systems are fully- or mostly-POSIX compliant.
+
+Windows is not natively POSIX compliant. However, some "modes" inside of Windows are, such as WSL
+(Windows Subsystem for Linux), gitbash, or some VSCode terminals.
+:::
+
+### Windows executables
+
+If your Windows machine has Python registered as the default application associated with `.py` files, then any Python
+scripts can be run as commands. However, only one Python can be registered at a time, so all Python scripts run this
+way will use the same Python environment. While all Python files should end in a `.py`, this naming is necessary for
+Windows to associate the file with Python, as opposed to on Linux where `.py` is a convention and the shebang associates
+the file with Python.
+
+Additionally, most Windows Python installs come with the [Python Launcher](https://docs.python.org/3/using/windows.html#python-launcher-for-windows)
+which, in addition to allowing specifying the version of Python, can also read shebang lines and emulate some of that behavior.
+This allows for shebang lines to be re-used between Linux, macOS, and Windows systems. However, on Windows the command must still
+be prefaced with another command (`py`).
+
+```bash
+py my_program.py
+# ✨ Hello from Python ✨
+```
+
+:::{tip}
+While there is no in-source format that can tell Windows what to do with a Python code file, executing a
+Python file with a shebang on Windows also does not cause any issues. Python just sees the whole line as
+a comment and ignores it! So even if you develop on Windows, it may be a good idea to add the shebang as
+the first line of your scripts, so that colleagues on different systems can also run it.
+:::
+
+## Executable comparisons
+
+### Pros of passing a file to `python`:
+- don't need execute permissions
+- works for every system
+- explicit about what you expect to happen
+
+### Pros of inserting a shebang to the file:
+- file is associated with specific python
+  - don't have to remember which
+- don't have to use the `python` command
+  - don't have to even remember it is a Python script
+
+
+# Share Code
```

Comment:
I think myst is going to complain about two `#` headings in a single doc, 

could ToC structure be something like

```
# Executing code
## Executing Scripts
### ...
## Executing Modules
### ...
## Entrypoints - `project.scripts`
### ...
```

or whatever order - i think that order has a nice build to it: "ok we are running single files, probably familiar, now we're running a file but it's a special file `__main__` (or `__name__ == "__main__"` block), and now we're running a module anywhere in our shell" but up 2 u.

Reply:
expanded on this a bit in next comment

Reply:
I mentioned this in a separate discussion about the executable lesson, but I originally saw this as two separate lessons, but I just drafted it as a single file. I though it was maybe too much to do all together. Also, the first section requires almost no other preliminary knowledge, while the second gains a lot by the audience at least already being familiar with packaging and package structure. It could instead be a single lesson, but I would structure it slightly differently.

Reply:
up to you! always in favor of smaller chunks :)

Reply:
I split it all the way in my next draft.